File: 2020/02/rust/src/main.rs
use regex::Regex;

fn main() {
    let input_file_content = std::fs::read_to_string(r"..\input.txt").unwrap();

    let pattern = Regex::new(r"(\d+)-(\d+) (\w+): (.+)").unwrap();

    let (mut counter1, mut counter2) = (0, 0);
    for line in input_file_content.lines(){
        let matches = pattern.captures(&line).unwrap();
        let low: usize = matches[1].parse().unwrap();
        let high: usize = matches[2].parse().unwrap();
        let character = matches[3].chars().next().unwrap();
        let code = &matches[4];
        let character_count = code.chars().filter(|&c| c == character).count();
        if (low..=high).contains(&character_count) {
            counter1 += 1;
        };
        if (code.chars().nth(low - 1) == Some(character)) != 
           (code.chars().nth(high - 1) == Some(character)) {
            counter2 += 1;
        }
    }
    println!("{}", counter1);
    println!("{}", counter2);
}

File: 2020/02/java/src/main/java/Main.java
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.regex.Pattern;

import io.vavr.collection.Stream;

public class Main {

    public static void main(String[] args) throws IOException {
        var lines = Stream.ofAll(
                Files.lines(Path.of("..", "input.txt")));
        var pattern = Pattern.compile("(\\d+)-(\\d+) (\\w+): (.+)");

        int counter1 = 0;
        int counter2 = 0;
        for (String line : lines) {
            var matcher = pattern.matcher(line);
            matcher.matches();
            var low = Integer.parseInt(matcher.group(1));
            var high = Integer.parseInt(matcher.group(2));
            var character = matcher.group(3).charAt(0);
            var code = matcher.group(4);
            var characterCount = code
                    .chars()
                    .filter(c -> c == character)
                    .count();
            if (characterCount <= high && characterCount >= low)
                counter1 += 1;
            if ((code.charAt(low - 1) == character) !=
                    (code.charAt(high - 1) == character))
                counter2 += 1;
        }

        System.out.println(counter1);
        System.out.println(counter2);
    }
}